documentation/matlab-api.md

```diff
@@ -0,0 +1,381 @@
+# MATLAB API discussion
+
+This document is intended to describe the means by which argument flags should be passed into PACE MATLAB programs.
+This is discussed in conjunction with an observation that a migration towards a Python interface is intended in the near future.
+
+## Current Use in Language
+### MATLAB
+MATLAB uses a mixture of flag and keyword value args. Examples from the standard language include, which mostly seem to revolve around interactions with the underlying system:
+```
+load(filename,'-ascii')
+whos('-file','durer.mat')
+checkcode('-id', filename)
+```
+
+Although Herbert's variant of `matlab_xunit` based `runtests` uses flags:
+```
+runtests(tests, '-verbose')
+```
+Core MATLAB seems to have made a move more towards keyword-value
+```
+arrayfun(@()(), list, 'UniformOutput', true)
+runtests(pwd,'IncludeSubfolders',true)
+runtests(tests, 'LoggingLevel', matlab.unittest.Verbosity.VERBOSE)
+```
+
+### Python
+Python's main method of handling variable arguments is to use the standard `*args` and `**kwargs` arguments.
+
+```
+def function_name(*args):
+    for arg in args:
+        print(arg)
+
+function_name(1,2,3,4)
+
+def function_name(**kwargs):
+    if kwargs.get('header', False):
+        print(header)
+        del kwargs['header']
+    for arg in kwargs.items():
+        print(arg)
+
+function_name(a=1, header='Hello World', b=4)
+```
+Any Python argument can be treated like a keyword argument.
+```
+def function_name(a,b,c):
+    print(a, b, c)
+
+function_name(1,c=2,b=1) # A is positional
+```
+
+Python also allows default (optional) arguments
+```
+def function_name(a=17):
+    print(a)
+
+function_name(31) # 31
+function_name()   # 17
+```
+
+Any or all of these can be used in combination with each other.
+
+### MATLAB Argument Parsers
+
+### `inputParser`
+Built-in `inputParser` supports positional, optional and required or optional key-value arguments
+
+```
+   defaultHeight = 1;
+   defaultUnits = 'inches';
+   defaultShape = 'rectangle';
+   expectedShapes = {'square','rectangle','parallelogram'};
+
+   p = inputParser;
+   validScalarPosNum = @(x) isnumeric(x) && isscalar(x) && (x > 0);
+   addRequired(p,'width',validScalarPosNum);
+   addOptional(p,'height',defaultHeight,validScalarPosNum);
+   addParameter(p,'units',defaultUnits,@isstring);
+   addParameter(p,'shape',defaultShape,...
+                 @(x) any(validatestring(x,expectedShapes)));
+   parse(p,width,varargin{:});
+```
+
+
+#### Pros
+- Very flexible
+- Well documented with potential online help
+- Standard MATLAB method
+- Explicit argument status
+- Keyword validation
+
+#### Cons
+- May change in future
+- Not extensible if doesn't meet requirements
```

Comment:
CON: there is a known issue that an optional character input followed by a keyword-value pair will not give what you think. This combination should be explicitly avoided.